### PR TITLE
Streaming ingest from blob storage

### PR DESCRIPTION
**Pull request by ohadbitt**
Streaming client and Managed streaming client support streaming ingest from storage

### CODE REVIEW

File: data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java [hunk 1]
```diff
@@ -95,6 +95,57 @@
         return execute(DEFAULT_DATABASE_NAME, command);
     }
 
+    @Override
+    public KustoOperationResult executeStreamingIngestFromBlob(String databaseName, String tableName, String blobUrl, ClientRequestProperties properties,
```
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java:99]: Why is it a separate implementation from executeStreamingIngest? they're nearly identical. Combine as much as you can to one function

File: data/src/main/java/com/microsoft/azure/kusto/data/Utils.java [hunk 2]
```diff
@@ -67,13 +67,14 @@
         // Hide constructor, as this is a static utility class
     }
 
-    static String post(CloseableHttpClient httpClient, String urlStr, String payload, InputStream stream, long timeoutMs, Map<String, String> headers,
+    static String post(CloseableHttpClient httpClient, String urlStr, String payload, InputStream stream, String blobUrl, long timeoutMs,
```
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/Utils.java:70]: This is getting messy. Maybe this should accept RequestEntity already?

File: data/src/main/java/com/microsoft/azure/kusto/data/Utils.java [hunk 3]
```diff
@@ -243,12 +244,19 @@
...
-    private static HttpPost setupHttpPostRequest(URI uri, String payload, InputStream stream, Map<String, String> headers) {
+    private static HttpPost setupHttpPostRequest(URI uri, String payload, InputStream stream, String blobUrl, Map<String, String> headers) {
         HttpPost request = new HttpPost(uri);
 
         // Request parameters and other properties. We use UncloseableStream to prevent HttpClient From closing it
-        HttpEntity requestEntity = (stream == null) ? new StringEntity(payload, ContentType.APPLICATION_JSON)
-                : new InputStreamEntity(new UncloseableStream(stream));
+        HttpEntity requestEntity;
+        if (blobUrl != null) {
+            requestEntity = new StringEntity(new IngestionSourceStorage(blobUrl).toString(), ContentType.APPLICATION_JSON);
```
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/Utils.java:253]: What is happening here? why do we need the wrapper class?

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java [hunk 4]
```diff
@@ -206,7 +210,6 @@
     /**
      * {@inheritDoc}
      * <p>
-     * This method behaves differently from the rest for {@link ManagedStreamingIngestClient} - since a blob already exists it makes more sense to enqueue it rather than downloading and streaming it, thus ManagedStreamingIngestClient skips the streaming retries and sends it directly to the queued client.</p>
      */
     @Override
     public IngestionResult ingestFromBlob(BlobSourceInfo blobSourceInfo, IngestionProperties ingestionProperties)
```
Comment (AsafMah) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java:215]: Again - this is a long & complex method that shares a lot of the code with ingest from stream.

Try to combine & resuse as much as possible

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/QueuedIngestClientImpl.java [hunk 5]
```diff
@@ -226,7 +226,7 @@ public IngestionResult ingestFromStream(StreamSourceInfo streamSourceInfo, Inges
                     shouldCompress);
             String blobPath = container.getContainer().getBlobContainerUrl() + "/" + blobName + container.getSas();
             BlobSourceInfo blobSourceInfo = new BlobSourceInfo(
-                    blobPath, 0); // TODO: check if we can get the rawDataSize locally - maybe add a countingStream
+                    blobPath, 0, streamSourceInfo.getSourceId()); // TODO: check if we can get the rawDataSize locally - maybe add a countingStream
```
Comment (AsafMah) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/QueuedIngestClientImpl.java:229]: What does this do? did we send it wrong so far?

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/StreamingIngestClient.java [hunk 6]
```diff
@@ -203,7 +202,10 @@
         return inputStream;
     }
 
-    IngestionResult ingestFromBlob(BlobSourceInfo blobSourceInfo, IngestionProperties ingestionProperties, BlobClient cloudBlockBlob)
+    IngestionResult ingestFromBlob(BlobSourceInfo blobSourceInfo,
```
Comment (AsafMah) [inline ingest/src/main/java/com/microsoft/azure/kusto/ingest/StreamingIngestClient.java:205]: This is a breaking change

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClientTest.java [hunk 7]
```diff
@@ -541,6 +543,18 @@ void IngestFromStream_IngestOverFileLimit_QueuedFallback(boolean leaveOpen, bool
         }
     }
 
+    @Test
+    void IngestFromBlob_IngestOverBlobLimit_QueuedFallback() throws Exception {
+        ingestionProperties.setDataFormat(IngestionProperties.DataFormat.JSON);
+
+        BlobSourceInfo blobSourceInfo = new BlobSourceInfo(
+                "https://blobPath.blob.core.windows.net/container/blob", MAX_STREAMING_SIZE_BYTES + 1);
```
Comment (AsafMah) [inline ingest/src/test/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClientTest.java:551]: Should we have an E2E Test?

**Review by AsafMah — CHANGES_REQUESTED** [review]

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/QueuedIngestClientImpl.java (hunk 5 above)
Comment (ohadbitt) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/QueuedIngestClientImpl.java:229]: yes

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/StreamingIngestClient.java (hunk 6 above)
Comment (ohadbitt) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/StreamingIngestClient.java:205]: its package private

File: ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java (hunk 4 above)
Comment (ohadbitt) [reply ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java:215]: Agreed with you there - disagree here . Not much in common

File: data/src/main/java/com/microsoft/azure/kusto/data/Utils.java [hunk 8]
```diff
@@ -67,13 +68,14 @@ private Utils() {
         // Hide constructor, as this is a static utility class
     }
 
-    static String post(CloseableHttpClient httpClient, String urlStr, String payload, InputStream stream, long timeoutMs, Map<String, String> headers,
-            boolean leaveOpen)
+    static String post(CloseableHttpClient httpClient, String urlStr, AbstractHttpEntity requestEntity, long timeoutMs,
+                       Map<String, String> headers,
+                       boolean leaveOpen)
```
Comment (AsafMah) [inline data/src/main/java/com/microsoft/azure/kusto/data/Utils.java:73]: Not yours but leaveOpen is unused

**Review by AsafMah — CHANGES_REQUESTED** [review]

File: data/src/main/java/com/microsoft/azure/kusto/data/Utils.java (hunk 8 above)
Comment (ohadbitt) [reply data/src/main/java/com/microsoft/azure/kusto/data/Utils.java:73]: it is - i just fixed it

File: ingest/src/test/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClientTest.java (hunk 7 above)
Comment (ohadbitt) [reply ingest/src/test/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClientTest.java:551]: added

**Pull request merged**